### ISSUE TITLE
Cancel in flight requests if another request is coming out

### ISSUE THREAD

**Issue by senadir** [woothemes/woocommerce#42640]
Issue reported by @manospsyx is when you have an in-flight request (update quantity from 3 to 4), and you click update again (4 to 5). Your result would jump back to 4 and then 5 again due to two requests coming back.

Given we have the most up to date state in the client, we want to cancel inflight requests in favor of the latest request.

https://user-images.githubusercontent.com/6165348/160151899-d1f2c800-c013-4129-a72d-451738a6d2e8.mov

**Comment by ankitguptaindia** [comment]
I am also facing this issue while working on Min/Max extension compatibility testing. More details are here in **Point-6** https://github.com/woocommerce/woocommerce-min-max-quantities/pull/152#issuecomment-1204989738. (Please feel free to drop a comment if you don't have access of that repo, I am happy to add details here)

**Comment by tarunvijwani** [comment]
Related (old) issue woocommerce/woocommerce#42650 

We also have the same issue on the Checkout block:

https://user-images.githubusercontent.com/11503784/185094907-00dac9af-b002-4819-8bce-6fae3e7b6008.mov

**Comment by senadir** [comment]
This can be generalized to:

- Selecting shipping rates.
- Updating Quantity.
- Possibility: typing an address.
And would remove all debounces.

**Comment by agung2001** [comment]
yeah also had the same issues, I think debouncing before doing API request will solve the issue

**Comment by senadir** [comment]
We do currently have debounce in place, increasing the amount of debounce will degrade the UX, ideally this issue should remove all usage of debounce.

**Comment by nickfmc** [comment]
This issue is causing major issues with plugins. for instance Advanced Coupons making it unusable when logged into the website and using Advanced Coupons you can't even change the quantity by one without it changing itself back..

_It is more apparent when our plugin is active because our plugin adds a validation on the cart for our cart condition feature. However, the root of the issue is still because the initial request to increase the quantity is not fully finished, when you increase the quantity again, it'll revert back to the initial request. So unless WooCommerce fixed this issue on the core WooCommerce first, we can't fix this issue._

**Comment by senadir** [comment]
Hey! I'm not really about the issue you're running into @nickfmc , can you provide a video or a live website to test on? Quantity will only be sent to the server after 400ms of inactivity, if you're clicking manually, it will continue to wait for you.

**Comment by nickfmc** [comment]
@senadir  this was a video the plugin dev made showing this issue, saying this is why their plugin is doing what it does. and the above comment was their reply to me explaining that with the plugin you can't change quantity at all it jumps right back.  Also logged out users work fine, but once a user is logged in the issue happens

https://www.loom.com/share/57541c6f67a24cebb2e9b6bbd9085f22

**Comment by ServiceWebLG** [comment]
Hello,

Is this issue with WooCommerce functionality resolved or in progress? I am encountering the same problem when using the Advanced Coupon plugin, which completely prevents users from using the quantity buttons (the quantity always resets to the initial request).

**Comment by pierorocca** [comment]
Reviewed and confirmed this is still valid. Marking as ready for engineering.

Still replicable on cart. It's no a great experience that's relying on smooth user input to deliver a smooth experience. Mini-cart has the opposite problem. While there's no racing condition, burst changes to quantity takes 3-4 seconds to resolve. Different issue that's almost as bad.

Claude suggests a combo of 1 + 4:

**1. Abort the in-flight request when a new one is issued (AbortController)**

Before dispatching a new request, abort the pending one. Handle the `AbortError` so it doesn't surface as an unhandled rejection:

```ts
const controllerRef = useRef<AbortController | null>(null);

const updateQuantity = (qty: number) => {
    controllerRef.current?.abort();
    controllerRef.current = new AbortController();
    apiFetch({ signal: controllerRef.current.signal, ... })
        .catch(err => {
            if (err.name === 'AbortError') return;
            // handle real errors
        });
};
```

The aborted response is ignored; only the last request settles UI state.

---

**2. Request serialisation (queue, don't race)**

Queue updates and send the next only after the previous resolves. Swallow errors on the chain so a single failed request doesn't freeze all future updates:

```ts
const pendingRef = useRef<Promise<void>>(Promise.resolve());

const updateQuantity = (qty: number) => {
    pendingRef.current = pendingRef.current
        .then(() => sendUpdate(qty))
        .catch(() => {}); // keep chain alive on failure
};
```

Adds latency — each click waits for the previous round-trip. Only viable if intermediate values are collapsed (see #3).

---

**3. Collapse intermediate values**

Keep the debounce but track "last desired value." On flush, send only the current intended quantity regardless of how many clicks happened during the wait. The in-flight guard (see #4) is still needed to prevent a slow response from overwriting a value the user has since changed.

---

**4. Optimistic UI with sequence gating**

Apply the new quantity locally immediately, then ignore server responses older than the last user action. Capture the previous value so a rejection can revert:

```ts
const seqRef = useRef(0);

const updateQuantity = (qty: number) => {
    const seq = ++seqRef.current;
    const previousQty = localQty;
    setLocalQty(qty); // optimistic
    sendUpdate(qty)
        .then(serverQty => {
            if (seq === seqRef.current) setLocalQty(serverQty);
        })
        .catch(() => {
            if (seq === seqRef.current) setLocalQty(previousQty); // revert on failure
        });
};
```

---

**5. Single "pending mutation" slot**

Track one pending update. When a new one arrives, mark the previous as superseded. On resolution, check the flag — if set, fire another request with the latest value; otherwise commit. In Redux, implement this by storing a `latestRequestId` in the slice and comparing it against the `requestId` in fulfilled/rejected action handlers — not via `createAsyncThunk`'s `condition` option, which only gates whether a thunk starts, not whether its result should be applied.

---

**In WooCommerce's context**, the right answer is **#1 (AbortController) + #4 (sequence gating)** as a safety net. Aborting frees the network slot immediately; sequence gating ensures that if a browser-buffered response from an aborted request somehow still resolves, it's discarded. The debounce stays, but its role narrows to collapsing fast bursts — after the burst, exactly one request fires and orphaned responses are safely ignored.
